workflow-manager/task/task.go

```diff
@@ -0,0 +1,156 @@
+	// Time is the timestamp on the batch, in UTC, with minute precision,
+	// formatted like "2006/01/02/15/04"
+	Date string `json:"date"`
+}
+
+func (i IntakeBatch) MarshalToJson() ([]byte, error) {
+	return json.Marshal(i)
+}
+
+// Enqueuer allows enqueuing tasks.
+type Enqueuer interface {
+	// Enqueue enqueues a task to be executed later. The provided completion
+	// function will be invoked once the task is either successfully enqueued or
+	// some unretryable error has occurred. A call to Stop() will not return
+	// until completion functions passed to any and all calls to Enqueue() have
+	// returned.
+	Enqueue(task Task, completion func(error))
+	// Stop blocks until all tasks passed to Enqueue() have been enqueued in the
+	// underlying system, and all completion functions pased to Enqueue() have
+	// returned, and so it is safe to exit the program without losing any tasks.
+	Stop()
+}
+
+// CreatePubSubTopic creates a PubSub topic with the provided ID. Returns error
+// on failure.
+func CreatePubSubTopic(ctx context.Context, project string, topicID string) error {
+	client, err := pubsub.NewClient(ctx, project)
+	if err != nil {
+		return fmt.Errorf("pubsub.newClient: %w", err)
+	}
+
+	if _, err := client.CreateTopic(ctx, topicID); err != nil {
+		return fmt.Errorf("pubsub.CreateTopic: %w", err)
+	}
+
+	return nil
+}
+
+// GCPPubSubEnqueuer implements Enqueuer using GCP PubSub
+type GCPPubSubEnqueuer struct {
+	topic     *pubsub.Topic
+	context   context.Context
+	waitGroup sync.WaitGroup
+	dryRun    bool
+}
+
+// NewGCPPubSubEnqueuer creates a task enqueuer for a given project and topic
+// in GCP PubSub. If dryRun is true, no tasks will actually be enqueued. Clients
+// should re-use a single instance as much as possible to enable batching of
+// publish requests.
+func NewGCPPubSubEnqueuer(ctx context.Context, project string, topicID string, dryRun bool) (*GCPPubSubEnqueuer, error) {
+	client, err := pubsub.NewClient(ctx, project)
+	if err != nil {
+		return nil, fmt.Errorf("pubsub.NewClient: %w", err)
+	}
+
+	return &GCPPubSubEnqueuer{
+		topic:   client.Topic(topicID),
+		context: ctx,
+		dryRun:  dryRun,
+	}, nil
+}
+
+func (e *GCPPubSubEnqueuer) Enqueue(task Task, completion func(error)) {
+	jsonTask, err := task.MarshalToJson()
+	if err != nil {
+		go completion(fmt.Errorf("marshaling task to JSON: %w", err))
+		return
+	}
+
+	if e.dryRun {
+		log.Printf("dry run, not enqueuing task")
+		e.waitGroup.Add(1)
+		go func() {
+			completion(nil)
+			e.waitGroup.Done()
+		}()
+		return
+	}
+
+	// Publish() returns immediately, giving us a handle to the result that we
+	// can block on to see if publishing succeeded. The PubSub client
+	// automatically retries for us, so we just keep the handle so the caller
+	// can do whatever they need to after successful publication and we can
+	// block in Stop() until all tasks have been enqueued
+	res := e.topic.Publish(e.context, &pubsub.Message{Data: jsonTask})
```

Comment:
Why don't you put the entire work (Publish + waiting for response) into the goroutine?

Reply:
You're right, it makes the error handling much simpler to do everything in the goroutine.